### ISSUE TITLE
C++11 std::condition_variable::wait_until() has very coarse timing accuracy (multiple of 3s)

### ISSUE THREAD

**Issue by ksdhans** [sba1/adtools#74]
I've been experimenting with std::condition_variable::wait_until(), and it seems to wait a multiple of about 3 seconds. So, it'll wait a minimum of 3 seconds, even if the wait_until time is just a few milliseconds away. This is with the machine almost at idle.

While the C++ specification doesn't guarantee any kind of clock granularity, 3 seconds is rather large. Plus, it raises questions about how efficiently this functionality has been implemented. If it were implemented using the timer.device, then I'd expect timing to be within milliseconds if the machine is almost at idle.

**Comment by ksdhans** [comment]
Forgot to say, this is with gcc 8.3.0

**Comment by ksdhans** [comment]
Here's some test code:

#include <chrono>
#include <thread>
#include <mutex>
#include <condition_variable>
#include <cstdio>

void testThread() {
	std::mutex mutex;
	std::condition_variable condition;
	printf("starting\n");
	std::unique_lock<std::mutex> mutexLock(mutex);
	printf("Entering loop\n");
	for(int i = 0; i < 30; ++i) {
		condition.wait_for(mutexLock, std::chrono::milliseconds(100));
		printf("Waited 0.1s, %d\n", i);
	}
}

int main(int argc, const char **argv) {
	std::thread thread(testThread);
	thread.join();
}


It should print a line every 0.1s, but on AmigaOS it regularly has long pauses between printing lines.

**Comment by ksdhans** [comment]
I've encountered this issue yet again. It would be great to finally have it fixed. 

Looking at the gthread code, it *is* using the timer.device, so the coarseness can't be from there. I spotted this comment:
/* Some timer arithmetics. TODO: Verify this! */

The following may have an error:
  if (tv_now.tv_sec + borrow <= (uint32_t)(abs_timeout->seconds))
    tv_now.tv_sec = abs_timeout->seconds - tv_now.tv_sec + borrow;
Shouldn't the last line have brackets around (tv_now.tv_sec + borrow)? It looks to me like an extra second is added here, on top of the second added to tv_now.tv_usec.
